crypto/x509/x509_req.c

```diff
@@ -286,6 +286,22 @@ void X509_REQ_get0_signature(const X509_REQ *req, const ASN1_BIT_STRING **psig,
         *palg = &req->sig_alg;
 }
 
+void X509_REQ_set0_signature(X509_REQ *req, ASN1_BIT_STRING *psig)
+{
+    if (req->signature)
+           ASN1_BIT_STRING_free(req->signature);
+    req->signature = psig;
+}
+
+void X509_REQ_set1_signature(X509_REQ *req, X509_ALGOR *palg)
+{
+    if (req->sig_alg.algorithm)
+        ASN1_OBJECT_free(req->sig_alg.algorithm);
+    if (req->sig_alg.parameter)
+        ASN1_TYPE_free(req->sig_alg.parameter);
+    req->sig_alg = *palg;
```

Comment:
This is still not quite right. It does not duplicate the internal pointers correctly. *Possibly* something like this might do it:

````
tmp = X509_ALGOR_dup(palg);
if (tmp == NULL)
    return 0;
req->sig_alg = *tmp;
OPENSSL_free(tmp);
````

But even this code makes all sorts of assumptions about the internal working of X509_ALGOR that it probably shouldn't. A nicer answer would be to add an internal `x509_algor_copy` function in x_algor.c that knows how to copy X509_ALGOR structures properly, e.g.

````
int x509_algor_copy(X509_ALGOR *dest, X509_ALGOR *src);
````

And then this setter simply becomes:
````
    return x509_algor_copy(&req->sig_alg, palg);
````

Reply:
Yes -fully agreed. So it seems we have two approaches in the codebase;

First -- much like all the other set0's which will do a free and the update the pointer; but not DUP (transfer mem mngt). E.g:

```
void NAMING_AUTHORITY_set0_authorityId(NAMING_AUTHORITY *n, ASN1_OBJECT* id)
{
    ASN1_OBJECT_free(n->namingAuthorityId);
    n->namingAuthorityId = id;
}
```

i.e. the recent setters.

Or - like what you allude to - and much like X509_NAME_set() and most other 'non 0/1'  plain XX_set()s which mostly do a DUP  -- the older setters. E.g:

```
int X509_NAME_set(X509_NAME **xn, const X509_NAME *name)
{
    X509_NAME *name_copy;

    if (*xn == name)
        return *xn != NULL;
    if ((name_copy = X509_NAME_dup(name)) == NULL)
        return 0;
    X509_NAME_free(*xn);
    *xn = name_copy;
    return 1;
}
```
So I can stay close the the other set0's and just prevent a leak (and note in the docs that the memory should not be freed - like the other set0's do).

Or do it more like the FOO_set_bar()'s which have a decent dup.

Reply:
And obviously use the x509_algor_copy() as that is shorted.

Reply:
And change the set0 to return an int (like set() - but not like the the void set0/1's).

Reply:
I think you're going to need to do the "set1" style approach for the X509_ALGOR. The thing is *this* setter is a little different to lots of the other setters in the codebase. The problem is for most other setters all we have to do is set a pointer. Easy. So in the `NAMING_AUTHORITY` example you give the members are all pointers:

````
struct NamingAuthority_st {
    ASN1_OBJECT* namingAuthorityId;
    ASN1_IA5STRING* namingAuthorityUrl;
    ASN1_STRING* namingAuthorityText;          /* i.e. DIRECTORYSTRING */
};
````
So, in a set 0, you can simply set the pointer. Or in a set1 you can dup it and then set it.

The sig_alg member of the X509_REQ structure is different though:

````
struct X509_req_st {
    X509_REQ_INFO req_info;     /* signed certificate request data */
    X509_ALGOR sig_alg;         /* signature algorithm */
    ASN1_BIT_STRING *signature; /* signature */
    CRYPTO_REF_COUNT references;
    CRYPTO_RWLOCK *lock;
# ifndef OPENSSL_NO_SM2
    ASN1_OCTET_STRING *sm2_id;
# endif
};
````
In this case the `sig_alg` member is *not* a pointer. It is an embeded structure. So you cannot simply do the set0 approach and set a pointer. You could individually set all the members and then free the passed in pointer - but I think this would be quite unexpected for the caller. Hence I think the set1 approach is the way to go. In this case you have dup all the *members* of the passed in X509_ALGOR, but not dup the whole containing X509_ALGOR.

This is why I suggest creating the function x509_algor_copy() to help implement this.

Reply:
Ok - understood and think I got it implemented right.